sql/addons_aggregates_v3.sql

```diff
@@ -0,0 +1,63 @@
+WITH
+  extended_addons AS (
+  SELECT
+  submission_date_s3 as submission_date,
+    sample_id,
+    client_id,
+    normalized_channel,
+    SUBSTR(app_version, 1, 2) AS app_version,
+    locale,
+    _addon.element.addon_id AS addon_id,
+    IF(_addon.element.foreign_install,
+      1,
+      0) AS is_foreign_install,
+    IF(_addon.element.is_system,
+      1,
+      0) AS is_system,
+    IF(_addon.element.is_web_extension,
+      1,
+      0) AS is_web_extension,
+    _addon.element.install_day AS install_day,
+    IF(_addon.element.addon_id IS NOT NULL
+      AND NOT _addon.element.is_system
+      AND NOT _addon.element.foreign_install
+      AND NOT _addon.element.addon_id LIKE '%mozilla%'
+      AND NOT _addon.element.addon_id LIKE '%cliqz%'
+      AND NOT _addon.element.addon_id LIKE '%@unified-urlbar%',
+      1,
+      0) AS is_self_install,
+    IF(_addon.element.addon_id LIKE "%@shield.mozilla%",
+      1,
+      0) AS is_shield_addon,
+    profile_creation_date
+  FROM
+    telemetry.main_summary_v4,
+    UNNEST(active_addons.list) AS _addon
```

Comment:
unnesting here and doing a group by at the end is a fairly innefficient way to do this in bigquery, compared to using window functions, e.g.:

```sql
WITH
  sample AS (
  SELECT
    submission_date_s3 AS submission_date,
    sample_id,
    client_id,
    normalized_channel,
    SUBSTR(app_version, 1, 2) AS app_version,
    locale,
    active_addons
  ...),
  grouped AS (
  SELECT
    * EXCEPT (active_addons),
    ROW_NUMBER() OVER (PARTITION BY submission_date, client_id, normalized_channel, app_version, locale) AS _n,
    ARRAY_CONCAT_AGG(active_addons.list) OVER (PARTITION BY  --
      client_id, normalized_channel, app_version, locale --
      ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS _addons
  FROM
    sample)
SELECT
  * EXCEPT (_n,
    _addons),
  (
  SELECT
    AS STRUCT --
    COUNTIF(_addon.element.addon_id IS NOT NULL
      AND NOT _addon.element.is_system
      AND NOT _addon.element.foreign_install
      AND NOT _addon.element.addon_id LIKE '%mozilla%'
      AND NOT _addon.element.addon_id LIKE '%cliqz%'
      AND NOT _addon.element.addon_id LIKE '%@unified-urlbar%') AS n_self_installed_addons,
    COUNTIF(_addon.element.foreign_install) AS n_foreign_installed_addons,
    ...
  FROM
    UNNEST(_addons) AS _addon ).*
FROM
  grouped
WHERE
  _n = 1
```

Reply:
OK. Tell me if I understand this right -- the idea is to use `ROW_NUMBER()` to enumerate each result for the (submission_date, client_id, ...) tuple that was used in the spark version's `.groupBy`. The final `WHERE _n = 1` serves to select only one result for each, similar to the `.distinct()` call at https://github.com/mozilla/python_mozetl/blob/master/mozetl/addon_aggregates/addon_aggregates.py#L141 .

The `ARRAY_CONCAT_AGG` creates one big array from every array in the partition, and the `COUNTIF`/`MIN` aggregation is done over this single array in the final SELECT.

Reply:
Also, can we ignore `sample_id` since bigquery can do partitioning on the relevant data directly?

Reply:
> OK. Tell me if I understand this right -- the idea is to use `ROW_NUMBER()` to enumerate each result for the (submission_date, client_id, ...) tuple that was used in the spark version's `.groupBy`.

Correct

> The final `WHERE _n = 1` serves to select only one result for each

Correct

> , similar to the `.distinct()` call at [...]

Not quite. `ROW_NUMBER() ... AS _n` plus `WHERE _n = 1` has the approximately the same effect as `GROUP BY`.

The `.distinct()` you reference actually does the equivalent of making your `extended_addons` CTE use `SELECT DISTINCT` instead of `SELECT`, presumably so that each addon_id is only counted once per client. This table pre-dates the existence of `active_addons` in `clients_daily_v6`, so I think it's working around not having a daily aggregation of that column. As such, I believe we can simplify this significantly by using `clients_daily` as the source table. Before we do that I'm going to attempt to reach out to stakeholders about the use and expected content for this dataset.

Reply:
Confirmed, we can simplify this by using clients_daily, like so:

```sql
SELECT
  submission_date_s3 AS submission_date,
  sample_id,
  client_id,
  normalized_channel,
  SUBSTR(app_version, 1, 2) AS app_version,
  locale,
  (
  SELECT
    AS STRUCT --
    COUNTIF(_addon.element.addon_id IS NOT NULL
      AND NOT _addon.element.is_system
      AND NOT _addon.element.foreign_install
      AND NOT _addon.element.addon_id LIKE '%mozilla%'
      AND NOT _addon.element.addon_id LIKE '%cliqz%'
      AND NOT _addon.element.addon_id LIKE '%@unified-urlbar%') AS n_self_installed_addons,
    COUNTIF(_addon.element.foreign_install) AS n_foreign_installed_addons,
    ...
  FROM
    UNNEST(active_addons) AS _addon ).*
FROM
  clients_daily_v6
WHERE
  client_id IS NOT NULL
  AND submission_date_s3 = @submission_date
```

and then there's no need to use `GROUP BY` or `ROW_NUMBER()` because `clients_daily` has already done that.

Reply:
> Also, can we ignore `sample_id` since bigquery can do partitioning on the relevant data directly?

`sample_id` is non-trivial to calculate in bigquery (it uses CRC32 which isn't available without a javascript UDF), and the table is 5.7 TB so we may eventually add clustering by sample_id so that it doesn't have to cost $26 to scan the whole table when a 1% sample would suffice.

Reply:
Great. I've updated to take this approach.